src/zone.c

```diff
@@ -0,0 +1,191 @@
+ *
+ * Authors:
+ *   Mathijs Mohlmann
+ **/
+
+#define _BSD_SOURCE
+
+#include "collectd.h"
+#include "common.h"
+#include "plugin.h"
+
+#include <sys/types.h>
+#include <sys/vm_usage.h>
+#include <procfs.h>
+#include <zone.h>
+
+#include "utils_avltree.h"
+
+#define	MAX_PROCFS_PATH	40
+#define MAX_ZONE_NAME 20
+#define FRC2PCT(pp)(((float)(pp))/0x8000*100)
+
+typedef struct zone_stats {
+	ushort_t      pctcpu;
+	ushort_t      pctmem;
+} zone_stats_t;
+
+static long pagesize;
+
+static int
+zone_compare(const zoneid_t *a, const zoneid_t *b)
+{
+	if (*a == *b)
+		return(0);
+	if (*a < *b)
+		return(-1);
+	return(1);
+}
+
+static int
+zone_read_procfile(char *pidstr, char *file, void *buf, size_t bufsize)
+{
+	int fd;
+
+	char procfile[MAX_PROCFS_PATH];
+	(void)snprintf(procfile, MAX_PROCFS_PATH, "/proc/%s/%s", pidstr, file);
+	while ((fd = open(procfile, O_RDONLY)) == -1) {
+		if ((errno != EMFILE) || (errno != ENFILE)) {
+			return(1);
+		}
+	}
+
+	if (pread(fd, buf, bufsize, 0) != bufsize) {
+		close(fd);
+		return (1);
+	}
+	close(fd);
+	return (0);
+}
+
+static int
+zone_submit_value(char *zone, gauge_t value)
+{
+	value_list_t vl = VALUE_LIST_INIT;
+	value_t      values[1];
+
+	values[0].gauge = value;
+
+	vl.values = values;
+	vl.values_len = 1; /*STATIC_ARRAY_SIZE (values);*/
+	sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+	sstrncpy (vl.plugin, "zone", sizeof (vl.plugin));
+ 	sstrncpy (vl.type, "zone_cpu", sizeof (vl.type));
+ 	sstrncpy (vl.type_instance, zone, sizeof (vl.type_instance));
+
+	return(plugin_dispatch_values (&vl));
+}
+
+static zone_stats_t *
+zone_find_stats(c_avl_tree_t *tree, zoneid_t zoneid)
+{
+	zone_stats_t *ret = NULL;
+	zoneid_t     *key = NULL;
+
+	if (c_avl_get(tree, (void **)&zoneid, (void **)&ret)) {
+		if (!(ret = malloc(sizeof(zone_stats_t)))) {
+			WARNING("no memory");
+			return(NULL);
+		}
+		if (!(key = malloc(sizeof(zoneid_t)))) {
+			WARNING("no memory");
+			return(NULL);
+		}
+		*key = zoneid;
+		if (c_avl_insert(tree, key, ret)) {
+			WARNING("error inserting into tree");
+			return(NULL);
+		}
+	}
+	return(ret);
+}
+
+static void
+zone_submit_values(c_avl_tree_t *tree)
+{
+	char          zonename[MAX_ZONE_NAME];
+	zoneid_t     *zoneid = NULL;
+	zone_stats_t *stats  = NULL;
+
+	while (c_avl_pick (tree, (void **)&zoneid, (void **)&stats) == 0)
+	{
+		getzonenamebyid(*zoneid, zonename, MAX_ZONE_NAME-1);
+		zone_submit_value(zonename, (gauge_t)FRC2PCT(stats->pctcpu));
+		free(stats);
+		free(zoneid);
+	}
+	c_avl_destroy(tree);
+}
+
+static c_avl_tree_t *
+zone_scandir(DIR *procdir)
+{
+	char         *pidstr;
+	pid_t         pid;
+	dirent_t     *direntp;
+	psinfo_t      psinfo;
+	c_avl_tree_t *tree;
+	zone_stats_t *stats;
+/*	size_t    physmem = sysconf(_SC_PHYS_PAGES) * pagesize;*/
+
+	if (!(tree=c_avl_create((int (*)
+				 (const void *, const void *))zone_compare))) {
+		WARNING("Failed to create tree");
+		return(NULL);
+	}
+
+	for (rewinddir(procdir); (direntp = readdir(procdir)); ) {
```

Comment:
`readdir()` returns `NULL` if all entries have been processed, so it is comparison and iteration at the same time.